File: 2_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        current = result
        carry = 0
        while l1 and l2:
            current.next = ListNode((l1.val + l2.val + carry) % 10)
            carry = (l1.val + l2.val + carry) // 10
            current = current.next

            l1 = l1.next
            l2 = l2.next

        if not l1 and not l2 and carry:
            current.next = ListNode(carry)

        if l1:
            l1 = self.addTwoNumbers(l1, ListNode(carry)) if carry else l1
            current.next = l1
        elif l2:
            l2 = self.addTwoNumbers(l2, ListNode(carry)) if carry else l2
            current.next = l2

        return result.next


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def __str__(self) -> str:
        return str(self.val) + ('->' + str(self.next))


a1 = ListNode(4)
# a2 = ListNode(9)
# a3 = ListNode(3)
# a4 = ListNode(1)
# a1.next = a2
# a2.next = a3
# a3.next = a4

b1 = ListNode(7)
b2 = ListNode(3)
# b3 = ListNode(4)
b1.next = b2
# b2.next = b3

s = Solution()
print(s.addTwoNumbers(a1, b1))

File: 1-100/2.add-two-numbers.js
/*
 * @lc app=leetcode id=2 lang=javascript
 *
 * [2] Add Two Numbers
 *
 * https://leetcode.com/problems/add-two-numbers/description/
 *
 * algorithms
 * Medium (31.98%)
 * Total Accepted:    1.1M
 * Total Submissions: 3.3M
 * Testcase Example:  '[2,4,3]\n[5,6,4]'
 *
 * You are given two non-empty linked lists representing two non-negative
 * integers. The digits are stored in reverse order and each of their nodes
 * contain a single digit. Add the two numbers and return it as a linked list.
 *
 * You may assume the two numbers do not contain any leading zero, except the
 * number 0 itself.
 *
 * Example:
 *
 *
 * Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 * Output: 7 -> 0 -> 8
 * Explanation: 342 + 465 = 807.
 *
 *
 */
/**
 * Definition for singly-linked list.
 */
function ListNode (val) {
  this.val = val
  this.next = null
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const addTwoNumbers = function (l1, l2) {
  let carry = 0
  const root = new ListNode()
  let p = root
  while (l1 || l2) {
    let sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry
    carry = Math.floor(sum / 10)
    p.next = new ListNode(sum % 10)
    p = p.next
    l1 = l1 ? l1.next : null
    l2 = l2 ? l2.next : null
  }

  if (carry) {
    p.next = new ListNode(1)
  }

  return root.next
}
